src/java/org/apache/cassandra/db/guardrails/Threshold.java

```diff
@@ -0,0 +1,282 @@
+    /**
+     * A facility for when the value to be guarded is built incrementally, but we want to abort the operation as soon as
+     * the abort threshold is reached, but only trigger the warning on the final value (and so only if the abort
+     * threshold hasn't also been reached).
+     * <p>
+     * Note that instances are neither thread safe nor reusable.
+     */
+    public class GuardedCounter
+    {
+        private final long warnValue;
+        private final long abortValue;
+        private final Supplier<String> what;
+
+        private long accumulated;
+
+        private GuardedCounter(Supplier<String> what, Config config)
+        {
+            // We capture the warn and abort value at the time of the counter construction to ensure we use
+            // stable value during the counter lifetime (and reading a final field is possibly at tad faster).
+            this.warnValue = warnValue(config);
+            this.abortValue = abortValue(config);
+            this.what = what;
+        }
+
+        /**
+         * The currently accumulated value of the counter.
+         */
+        public long get()
+        {
+            return accumulated;
+        }
+
+        /**
+         * Add the provided increment to the counter, aborting the operation if the counter after this addition crosses
+         * the abort threshold.
+         *
+         * @param increment the increment to add
+         */
+        public void add(long increment)
+        {
+            accumulated += increment;
+            if (accumulated > abortValue)
+                triggerAbort(accumulated, abortValue, what.get());
+        }
+
+        /**
+         * Trigger the warn if the currently accumulated counter value crosses warning threshold and the abort threshold
+         * has not been crossed yet.
+         * <p>
+         * This is generally meant to be called when the guarded value is complete.
+         *
+         * @return {@code true} and trigger a warning if the current counter value is greater than the warning
+         * threshold and lesser than or equal to the abort threshold, {@code false} otherwise.
+         */
+        public boolean checkAndTriggerWarning()
+        {
+            if (accumulated > warnValue && accumulated <= abortValue)
+            {
+                triggerWarn(accumulated, warnValue, what.get());
+                return true;
+            }
+            return false;
+        }
+    }
+
+    /**
+     * Configuration class containing the thresholds to be used to check if the guarded value should trigger a warning
+     * or abort the operation.
+     */
+    public interface Config
+    {
+        public static Config DEFAULT = new Config()
+        {
+            @Override
+            public long getWarnThreshold()
+            {
+                return -1;
+            }
+
+            @Override
+            public long getAbortThreshold()
+            {
+                return -1;
+            }
+        };
```

Comment:
deadcode (transitively, caller of caller is not used); can we remove?

Reply:
Removed, we'll probably add it back to the framework when have the use cases for it.